src/rnn/types.h

```diff
@@ -83,8 +89,24 @@ class States {
   States select(const std::vector<IndexType>& selIdx, // [beamIndex * activeBatchSize + batchIndex]
                 int beamSize, bool isBatchMajor) const {
     States selected;
+    Expr indices;
+    // I think this doesn't work if model split among gpus but not sure if it matters
+    
+    for (auto& state : states_) {
```

Comment:
Please add a comment what this logic does, as it is not obvious why the old code is not working.

Reply:
There is nothing wrong with the old code. The comment you highlighted is also wrong I think. I wrote this when I was much less familiar with Marian so I will just remove that comment.

Reply:
I added a comment in a recent commit but I'm not sure if it's clear enough. If so, feel free to resolve this.